Introducing bundled sub plugins

Jermolene opened the pull request:
# Introduction

This PR introduces "bundled sub-plugins" a new feature that allows plugins to contain further sub-plugins that are processed in the same way as top level plugins.

This change is intended to address several related needs:
 
* Shipping plugins with multiple translations, without having to package them as separate plugins (this PR is intended to fix #8435)
* Making plugins as separate sub-components that can be independently disabled and overridden by end users
* Shipping themes, languages and normal plugins together as a single container plugin with no dependencies

# Usage

To create a sub-plugin, add the plugin tiddler as a payload tiddler within the parent plugin. Any shadow tiddler with their type set to `application/json` and also has a `plugin-type` field will be processed as a sub-plugin.

Under Node.js, sub-plugins are created by placing a `plugin.info` file in a subdirectory of the parent plugin folder.

# Progress

* [x] Update plugin switching logic to accommodate sub-plugins
* [ ] Dynamic handling of changes to `$:/config/Plugins/Disabled/*` tiddlers
* [ ] Update control panel listing of installed plugins to indicate sub-plugins
* [ ] Filter operator to retrieve the sub-plugins of a plugin
* [ ] `is[subplugin]` filter operator
* [ ] Update the demo CodeMirror embedded translation to use real language strings. We should put a call out for translators
* [ ] Update control panel "add plugin" dialogue to show sub-plugins
* [ ] User documentation updates
* [ ] Documentation updates for wikitext authors
* [ ] Update the Translators plugin so that it can be used to create language sub-plugins as well as creating core translations

# Backwards Compatibility

The existing term "sub-plugin" refers to a plugin that is subordinate to a parent plugin. Currently, the only way to create a sub-plugins is to set their `parent-plugin` field to the title of their parent plugin. The plugin library listing displays sub-plugins hierarchically under an expandable button. The plugin library also automatically installs the parent plugin of sub-plugins (along with any plugins listed in the `dependents` field).

This PR will eventually add the hierarchical display to the listing of installed plugins.

There are no breaking changes to the internal JavaScript API.

# Subtleties and gotchas

* Sub-plugins cannot currently contain further sub-plugins.
* Disabling the parent plugin via the `$:/config/Plugins/Disabled/` mechanism always disables all the sub-plugins. If the parent plugin is enabled then the sub-plugins can be individually disabled via the same mechanism
* Sub-plugins with a given title are independent of any tiddler or shadow tiddler with the same title. In particular, that means that sub-plugins cannot be overridden by creating a tiddler of the same title.

# Open Questions

* Should we require plugins that contain sub-plugins to be specifically marked (eg by a special field)? It would avoid the problem whereby any existing plugins that contain plugins as payload tiddlers would unexpectedly be loaded.

Jermolene committed:
Initial Commit

Jermolene committed:
Refactor some symbols for clarity

All these symbols are inaccessible outside the wiki class closure, and so no backwards compatibility issues

Jermolene committed:
More refactoring for clarity

Jermolene committed:
Use "bundled" rather than "intrinsic"

Jermolene committed:
Basic loading of bundled subplugins

Very basic. Doesn't cope with any changes to the plugins after startup, and doesn't remove sub-plugins when the parent plugin is removed.

linonetwo commented:
> plugins that contain sub-plugins to be specifically marked

Don't need to, let plugin library do this. And show them as

> hierarchical display to the listing of installed plugins

---

But how will sub-plugins switch based on language?

Jermolene commented:
Thanks @linonetwo 

> Don't need to, let plugin library do this. And show them as

I think this is incomplete, but I don't think it addresses the backwards compatibility problem I mentioned.

> But how will sub-plugins switch based on language?

The sub-plugins will be marked as add-on language plugins, and will be automatically included when switching to the corresponding language. Similar to background palettes in #8702

Jermolene committed:
Merge branch 'master' into intrinsic-sub-plugins

Jermolene committed:
More efficient checks if a tiddler is a plugin

Jermolene committed:
Merge branch 'master' into intrinsic-sub-plugins

Jermolene committed:
Clarify comment

Jermolene committed:
Deal with dynamic loading and unloading of plugins

More enhancements needed for the test framework too in order to allow events to be processed manually, without relying on the nexttick handler which is awkward in Jasmine

Jermolene committed:
Move subplugin tests into their own folder

Jermolene committed:
Test for loading language plugins at startup

Jermolene committed:
Code typo

Jermolene committed:
Fix testcase export to include plugins

Jermolene committed:
Add test for dynamically loading a language plugin

Jermolene committed:
Remove debug code

Jermolene committed:
Cleaning up

pmario reviewed (commented):
Just some thoughts

Jermolene committed:
Add support for add-on language and theme plugins

Leilei332 committed:
Allow building external core version of main site (#8953)

Jermolene committed:
Merge branch 'tiddlywiki-com'

Jermolene committed:
Fix accidentally disabled test

Jermolene committed:
Add comment

Jermolene committed:
Add subplugins for codemirror languages

The French translation is just for demo purposes. It would be helpful to replace it with some real translations.

Jermolene committed:
Display subplugins in control panel listing

Jermolene commented on boot/boot.js:
```diff
@@ -1068,6 +1068,10 @@ $tw.Tiddler.prototype.hasField = function(field) {
 	return $tw.utils.hop(this.fields,field);
 };
 
+$tw.Tiddler.prototype.isPlugin = function() {
```
Good idea – I've added `is[subplugin]` in 9dadd3fafb2ac05c05712d41fa302a0ac169e170

Jermolene commented on boot/boot.js:
```diff
@@ -1353,102 +1357,133 @@ $tw.Wiki = function(options) {
+	};
+
 	// Get plugin info for a plugin
 	this.getPluginInfo = function(title) {
```
The full pluginInfo structure is enormous because it contains the text of all the plugins, and so I don't think it would make sense to provide it in full. But I'm struggling to think of applications; the internal pluginInfo structure is required to support the plugin mechanism, but I don't think there's anything there that isn't available via inspecting tiddlers in the usual way.

Jermolene commented:
Hi @linonetwo – some changes to this PR made earlier today to the CodeMirror plugin demonstrate how to bundle multiple languages within a plugin. I'm behind on updating the docs, but hopefully you can see the general idea of how subplugins are ordinary plugin tiddlers that are nested inside another plugin.

Jermolene commented:
I would appreciate some thoughts on naming. I don't think we can use "sub-plugin" because it is a term we already use, with a slightly different meaning.

I currently favour "nested plugin", but would welcome other suggestions.

Jermolene committed:
Merge branch 'master' into intrinsic-sub-plugins

joshuafontany commented:
I like nested or maybe bundled plugin. Really interesting concept.

linonetwo commented:
All plugins are "bundled" (while specifically they are "bundled tiddlers"), so nested is better. BTW, what are sub-plugins previously? Those added "parent" field I thought there were "children plugins".

And I see, the "language" plugin switcher instance will switch all nested plugins, no matter how deep they are:

![图片](https://github.com/user-attachments/assets/240dfebb-04c8-4e25-bd19-c21bda0b7014)

It could also be named "deep plugin" like deep web.

Jermolene commented:
Thank you @joshuafontany @linonetwo 

> All plugins are "bundled" (while specifically they are "bundled tiddlers"), so nested is better.

Thanks, I think there's sufficient consensus to switch to nested.

> BTW, what are sub-plugins previously? Those added "parent" field I thought there were "children plugins".

That is correct, these are plugins with a "parent-plugin" field. The docs refer to them as "sub-plugins" in a couple of places.

It's probably not too late to retcon a different meaning to "sub-plugin", but I think it's not a great name because it only conveys the most basic idea of hierarchy, where what we really need here is a word that carries the meaning of containment.

> But they don't show "nested" on control panel, insteand they feels "connected", because they are separated on control panel, and when I delete or disable one, all of them are deleted.

I am wondering about introducing categories/folders into plugin listings that would be expandable, and then using the `parent-plugin` field as one of the ways to create these categories.

> And I see, the "language" plugin switcher instance will switch all nested plugins.

That's right, it looks like ChatGPT has made a decent fist of following the code.

CrossEye commented:
This looks like a very good idea.

I'm curious about this:

> * Sub-plugins cannot currently contain further sub-plugins.

Is that an up-front design decision or something that falls out of the implementation?  Once we have nested plugins, I would expect recursive nesting.  (I'm not saying I need it, though!)

One other minor question... that may get cleared up when I have time to read the code:  While we know that the root wiki will shadow a tiddler from a plugin, is it also true that the plugin will shadow a tiddler from one of its nested plugins?

linonetwo commented:
> One other minor question... that may get cleared up when I have time to read the code: While we know that the root wiki will shadow a tiddler from a plugin, is it also true that the plugin will shadow a tiddler from one of its nested plugins?

But they don't show "nested" on control panel, insteand they feels "connected", because they are separated on control panel, and when I delete or disable one, all of them are deleted.

Jermolene commented:
> > * Sub-plugins cannot currently contain further sub-plugins.
> 
> Is that an up-front design decision or something that falls out of the implementation? Once we have nested plugins, I would expect recursive nesting. (I'm not saying I need it, though!)

Supporting recursive nesting would make the code a little more complex, and would have a modest performance impact because of the need to scan the tiddlers of each subplugin as well as scanning those of the top level plugins. I started out by not supporting recursive nesting because it's easier to add a feature in the future than to take away an ill advised one. If any use cases emerge then we can revisit the question.

> One other minor question... that may get cleared up when I have time to read the code: While we know that the root wiki will shadow a tiddler from a plugin, is it also true that the plugin will shadow a tiddler from one of its nested plugins?

Yes. The plugin unpacking process starts by extracting all the sub-plugins and then placing them in a list along with the top level plugins. From then on there is no difference between the two types of plugins. That means that you can use the `plugin-priority` field to arrange for a particular sub-plugin to be shadowed by the parent plugin or the other way around.

Jermolene committed:
Merge branch 'master' into intrinsic-sub-plugins

Jermolene commented:
> I personally do not want to include all available languages if I do import a core plugin.

The tradeoff here is the needs of people who care about the size of the plugins that they use, and the needs of end users who happen to need a language other than English. For those users, navigating the options would be likely to be confusing.

A more philosophical point is that we value self-containedness. When things are self-contained, you don’t need to worry about coordinating the independent parts, you’ve always got everything you might need. A tool packaged as a single file is a lot easier to handle than managing the dependency and version matching between different versions.

Another consideration is that part of what we do is try to help and provide support for users. Having everybody standardise on the same thing makes it easier to provide that support. We see this in the way that being able to set up minimal test cases is so helpful for debugging problems.

Finally, I would like to understand exactly what problems are raised by a plugin being slightly larger than it needs to be. I don’t see a real issue for the majority of users.

linonetwo commented:
> ATM we can add language plugins with `parent-plugin` set.

That is true, but will require user to search for language pack, not working out-of-box. And too many language sub-plugin will pollute the CPL.
So I just hope this could be merged soon, even a simple version.
